### PR TITLE
chore(): v8 major release [WIP]

### PR DESCRIPTION
**Pull request by kamilmysliwiec**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Other... Please describe: Release
```

## What is the current behavior?

N/A

### Features

- **common**: refactor the built-in console logger class #6142 #5353, separate out `Logger` wrapper class and `ConsoleLogger` class, add buffering logs feature (buffer init logs and flush them out when a custom logger is assigned), add log level output #6172, support log levels feature nestjs/docs.nestjs.com#1613
- **common**: add `StreamableFile` class for streaming files from controller routes (cross-platform) https://github.com/nestjs/nest/pull/5278
- **common**: add `ParseFloatPipe` and `ParseEnumPipe` classes https://github.com/nestjs/nest/pull/6232
- **core**: add lazy module loader class https://github.com/nestjs/nest/pull/6386
- **core**: add `RouterModule` for defining per-module controller prefixes 
- **core**: add the ability to exclude certain routes from the global prefix https://github.com/nestjs/nest/pull/5291
- **core**: API versioning feature https://github.com/nestjs/nest/pull/6278
- **microservices**: support multiple event subscribers (`@EventPattern()` can be applied to multiple methods) https://github.com/nestjs/nest/pull/6334
- **microservices**: support passing a property key to the `@Payload()` decorator https://github.com/nestjs/nest/pull/5975
- **websockets**: support passing a property key to the `@MessageBody()` decorator
- **websockets**: (`WsAdapter`) support registering gateways on different paths and let them share the same HTTP server https://github.com/nestjs/nest/pull/6321

### Improvements

- **core**: sort modules topologically when invoking lifecycle hooks https://github.com/nestjs/nest/pull/6236
- **core**: use class references as providers/controllers/injectables keys instead of their names 
- https://github.com/nestjs/nest/pull/6141 fixes https://github.com/nestjs/nest/issues/5591 and https://github.com/nestjs/nest/issues/789
- **core**: `BaseExceptionFilter` supports `http-errors` (and any other error objects that have the `statusCode` property specified) out of the box now https://github.com/nestjs/nest/pull/6349/commits/cdd51b002888e2d4f249c271a958f8e7e1381997
- **core**: set `Error.name` property to the name of the class (all exceptions that inherit from the built-in `HttpException`) https://github.com/nestjs/nest/pull/5859

### Bug fixes

- **common**: add a missing type argument to `mixin()` function https://github.com/nestjs/nest/pull/6216
- **core**: use `router.all()` method for routes annotated with the `@All()` decorator (instead of `router.use()`) https://github.com/nestjs/nest/pull/6237
- **core**: reject on server bind failures (`listen()`) https://github.com/nestjs/nest/pull/3360
- **websockets**: specify `@nestjs/platform-socket.io` as an optional peer dependency to support PnP mode https://github.com/nestjs/nest/pull/6217

### Dependencies

- **microservices**: update to use `@grpc/grpc-js` package (instead of `grpc`) https://github.com/nestjs/nest/pull/4806
- **microservices**: upgrade NATS to v2 🙌 
- **platform-socket.io**: upgrade Socket.io to v4 🙌 

### Deprecations

- **common**: deprecate `HttpModule` to remove hard dependency on the `axios` library. `HttpModule` is published as a new `@nestjs/axios` package now
- **core**: deprecate the `listenAsync()` (use `listen()`) and `startAllMicroservicesAsync()` (use `startAllMicroservices()`) methods
- **microservices**: deprecate the `listenAsync()` method (use `listen()` instead)

## Does this PR introduce a breaking change?
```
[x] Yes
[ ] No
```

## Other information

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
If anyone wants to test v8 with their applications, you can install `8.0.0-alpha.1` version now (tag "next"). Example: `npm i @nestjs/common@next` (make sure to install `@nestjs/core`, `@nestjs/platform-express`, `@nestjs/microservices` etc too) cc @jmcdo29 @BrunnerLivio 🙌

**Comment by erescobar** [comment]
Hi, I get this error when I try to run an app using alpha version

`Error: Nest cannot export a provider/module that is not a part of the currently processed module (ConfigModule). Please verify whether the exported ConfigHostModule is available in this particular context.`

**Comment by kamilmysliwiec** [comment]
@erescobar can you create an issue and link a minimum reproduction repository?

File: integration/microservices/e2e/sum-kafka.spec.ts [hunk 1]
```diff
@@ -14,7 +14,8 @@ describe('Kafka transport', function () {
   let app: INestApplication;
 
   // set timeout to be longer (especially for the after hook)
-  this.timeout(50000);
+  this.timeout(30000);
+  this.retries(10);
```
Comment (jmcdo29) [inline integration/microservices/e2e/sum-kafka.spec.ts:18]: I'm guessing this it to generally fix the kafka integration test errors we're getting?

Comment (kamilmysliwiec) [reply integration/microservices/e2e/sum-kafka.spec.ts:18]: correct @jmcdo29

**Comment by MickL** [comment]
`alpha.3` works like a charm for me. Couldnt find `@nestjs/axios` tho.

**Comment by lukadriel7** [comment]
Hello, Thank you for the great job.
I have a question about nats v2 integration, will it also support Jetstream or just  PubSub ?

**Comment by scorsi** [comment]
Looking forward for the official stable release for the nats 2.0 driver 👀 
Good job btw the alpha works fine !

**Comment by lukadriel7** [comment]
Hello, any chance on getting jetstream support on NATS v2 ?

**Comment by kamilmysliwiec** [comment]
@lukadriel7 there's no plan in using JetStreams by default, but you could always integrate them by creating a Custom Transport Strategy https://docs.nestjs.com/microservices/custom-transport

**Comment by lukadriel7** [comment]
> @lukadriel7 there's no plan in using JetStreams by default, but you could always integrate them by creating a Custom Transport Strategy https://docs.nestjs.com/microservices/custom-transport

OK got it. Thanks

**Comment by budnieswski** [comment]
The gRPC server must be initialized (`this.grpcClient.start()`) in callback of `bindAsync`, otherwise throw an error: "Stream refused by server"

**Comment by GuilhermeMLS** [comment]
> The gRPC server must be initialized (`this.grpcClient.start()`) in callback of `bindAsync`, otherwise throw an error: "Stream refused by server"

`bindAsync` is actually wrapped in a promise, it will not throw error.

**Comment by moehlone** [comment]
Hi and thanks for the work first! Is it roughly predictable when this release can be published?

**Comment by versNumbers** [comment]
> > Is it roughly predictable when this release can be published?
> 
> @moehlone [#6860 (comment)](https://github.com/nestjs/nest/issues/6860#issuecomment-816345096)


@micalevisk I think he is referring to: [#5676 (comment)](https://github.com/nestjs/nest/issues/5676#issuecomment-771613092)

**Comment by MickL** [comment]
Hey @kamilmysliwiec , after the release of `8.0.0-alpha.3` you integrated NATS 2.0. Could you please make an `alpha.4` release so we can use (and test) NATS 2.0?

**Comment by iamkhalidbashir** [comment]
> Hey @kamilmysliwiec , after the release of `8.0.0-alpha.3` you integrated NATS 2.0. Could you please make an `alpha.4` release so we can use (and test) NATS 2.0?

We are in a position to add NATS into our nestjs mega app, and are worried that we have to migrate from Nats 1 to NATS 2 later on
@kamilmysliwiec can we release alpha.4 ?

**Comment by kamilmysliwiec** [comment]
`alpha.4` is published now :)

**Comment by MickL** [comment]
Alpha 4 is working fine for me with Nats 2.x :) @kamilmysliwiec I see there are new logs (Nats Ping, Nats reconnect). Is it possible to configure this logging? E.g. I would like to not see the ping in the logs (over days and weeks it will spam the logs) and I want to see only one time reconnect (currently it spams one per second which would result in massive logs after 1 hour downtime).

**Comment by MickL** [comment]
@kamilmysliwiec Using the CacheModule throws `Cannot read property 'httpAdapter' of undefined`. Are there changes I've missed?

**Comment by gperdomor** [comment]
@kamilmysliwiec to use v8 is enough to change the version of the packages in package.json or we need to do something else?

**Comment by kamilmysliwiec** [comment]
@gperdomor there are a few minor breaking changes. You can find the migration guide & instructions on the `8.0.0` branch of the docs https://github.com/nestjs/docs.nestjs.com/tree/8.0.0

**Comment by manjufy** [comment]
> @gperdomor there are a few minor breaking changes. You can find the migration guide & instructions on the `8.0.0` branch of the docs https://github.com/nestjs/docs.nestjs.com/tree/8.0.0

This page https://github.com/nestjs/docs.nestjs.com/tree/8.0.0 seems to be broken.

Is there a reference guide I can refer to for move from v7.5 to v8

**Comment by wrslatz** [comment]
https://docs.nestjs.com/migration-guide

**Comment by wrslatz** [comment]
> Is there a reference guide I can refer to for move from v7.5 to v8

https://docs.nestjs.com/migration-guide

**Comment by manjufy** [comment]
> > Is there a reference guide I can refer to for move from v7.5 to v8
> 
> https://docs.nestjs.com/migration-guide

Thank you @wrslatz

**Pull request merged**